CMakeLists.txt

```diff
@@ -198,125 +206,208 @@ else()
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picoquic_ct PRIVATE loglib)
-
-add_executable(picohttp_ct picohttp_t/picohttp_t.c
-    ${PICOQUIC_TEST_LIBRARY_FILES}
-    ${PICOHTTP_TEST_LIBRARY_FILES}
-)
-
-target_link_libraries(picohttp_ct
-    picoquic-log
-    picoquic-core
-    picohttp-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picohttp_ct PRIVATE loglib picohttp)
-
-add_executable(picoquic_sample
-    sample/sample.c
-    sample/sample_client.c
-    sample/sample_server.c
-)
-
-target_link_libraries(picoquic_sample
-    picoquic-log
-    picoquic-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picoquic_sample PRIVATE loglib picoquic)
-
-set(TEST_EXES picoquic_ct picohttp_ct)
+    PRIVATE
+        ${PTLS_LIBRARIES}
+        ${OPENSSL_LIBRARIES}
+    PUBLIC
+        picoquic-core)
+
+target_include_directories(picohttp-core
+    PRIVATE
+        ${PTLS_INCLUDE_DIRS}
+        ${OPENSSL_INCLUDE_DIR}
+    PUBLIC
+        picoquic)
+
+set_picoquic_build_options(picohttp-core)
+
+include(CTest)
+
+if(BUILD_TESTING AND picoquic_BUILD_TESTS)
+
+    add_executable(picoquicdemo
+        picoquicfirst/picoquicdemo.c
+        picoquicfirst/getopt.c)
+
+    target_link_libraries(picoquicdemo
+        PRIVATE
+            ${PTLS_LIBRARIES}
+            ${OPENSSL_LIBRARIES}
+            #${CMAKE_DL_LIBS}
+            picoquic-log
+            picoquic-core
+            picohttp-core)
+
+    target_include_directories(picoquicdemo
+        PRIVATE
+            #${PTLS_INCLUDE_DIRS}
+            #${OPENSSL_INCLUDE_DIR}
+            #picoquic
+            #loglib
+            picohttp)
+
+    set_picoquic_build_options(picoquicdemo)
+
+    add_executable(picolog_t picolog/picolog.c)
+
+    target_link_libraries(picolog_t
+        PRIVATE
+            #${PTLS_LIBRARIES}
+            #${OPENSSL_LIBRARIES}
+            #${CMAKE_DL_LIBS}
+            picoquic-log
+            picoquic-core)
+
+    target_include_directories(picolog_t
+        PRIVATE
+            #picoquictest
+            loglib)
+
+    set_picoquic_build_options(picolog_t)
+
+        add_executable(picoquic_ct
+            picoquic_t/picoquic_t.c
+            ${PICOQUIC_TEST_LIBRARY_FILES})
+
+    target_link_libraries(picoquic_ct
+        PRIVATE
+            ${PTLS_LIBRARIES}
+            ${OPENSSL_LIBRARIES}
+            #${CMAKE_DL_LIBS}
+            picoquic-log
+            picoquic-core)
+
+    target_include_directories(picoquic_ct
+        PRIVATE
+            ${PTLS_INCLUDE_DIRS}
+            ${OPENSSL_INCLUDE_DIR}
+            #loglib
+            picoquictest)
+
+    set_picoquic_build_options(picoquic_ct)
+
+    add_executable(picohttp_ct
+        picohttp_t/picohttp_t.c
+        ${PICOQUIC_TEST_LIBRARY_FILES}
+        ${PICOHTTP_TEST_LIBRARY_FILES})
+
```

Comment:
We should replace " ${PICOQUIC_TEST_LIBRARY_FILES}" by a "picoquic-test" library, which would be compiled once, rather than compiling the test code twice.

Reply:
Good idea. I'll work on that.